### ISSUE TITLE
intercept consumes an excessive amount memory (>20GB building wine)

### ISSUE THREAD

**Issue by MetaDark** [rizsotto/Bear#349]
**Describe the bug**
I cannot generate a `compile_command.json` when building wine because intercept completely exhausts my available memory (16GB + 8GB swap).

Memory usage stays below 10GB until the wine build completes, and then rapidly increases. I assume this is related to serialization.

**To Reproduce**
1. Follow instructions for building wine: https://wiki.winehq.org/Building_Wine
2. Run make step with bear:
```shell
bear -- make -j $(nproc)
```

Alternatively use the nix package manager to reproduce my environment:
1. Install Nix package manager (see https://nixos.org/manual/nix/stable/#chap-installation):
```shell
curl -L https://nixos.org/nix/install | sh
```
2. Install bear & git:
```shell
nix-env -iA nixpkgs.bear nixpkgs.git
```
3. Clone wine:
```shell
git clone git://source.winehq.org/git/wine.git
cd wine
git checkout 221fdb09b4c20da45e9834aa0cae27dcc75ea27b # probably not necessary, but this is the commit I used
```
4. Setup temporary build shell for wine:
```shell
nix-shell -E 'with import <nixpkgs> {}; (winePackagesFor "wine64").full.override { wineRelease = "unstable"; }'
```
5. Run patch & configure phases
```shell
phases='patchPhase configurePhase' genericBuild
```
6. Run make with bear:
```shell
bear -- make -j $(nproc)
```

**Expected behavior**
Should be able to build with bear if the project can build without bear using less than half the memory.

**Environment:**
 - OS name: Linux
 - OS version: NixOS, 21.03.git.56bb1b0f7a3 (Okapi)
 - OS architecture: x86_64
 - Bear version: 3.0.3
 - Bear install method: from [nixpkgs](https://github.com/Atemu/nixpkgs/blob/56bb1b0f7a33e5d487dc2bf2e846794f4dcb4d01/pkgs/development/tools/build-managers/bear/default.nix). eg. `nix-env -iA nixpkgs.bear`

**Additional context**
 - Can you give us a reference to the project that you are running against this tool?
Wine: git://source.winehq.org/git/wine.git (specifically commit 221fdb09b4c20da45e9834aa0cae27dcc75ea27b)
 - What build tools this project is using?
GCC, autoconf, Make, ... (see https://wiki.winehq.org/Building_Wine)
 - What architecture you are trying to compile for?
x86_64
 - Could you attach build logs? If you can re-run the command which failed (e.g. `bear make`) with extra verbose log switches (e.g. `bear -vvvv make`) and send the output of it, that would help a lot.
  I will try another build to get this output.

**Comment by jbmaillet** [comment]
I do full Linux embedded systems and Android AOSP builds on a (16GB + 32GB swap) machine, and I can confirm that there has been a noticeable performance loss since at least 3.0.3. It still does work for me, but end up using most of the swap and of course taking forever to complete (30 minutes post-build?). Parallel builds have become a non-option for my use cases. As a workaround, for some situations I switched back to a Debian docker image with bear 2.4.3.

**Comment by rizsotto** [comment]
Thanks folks for the report. I'm working on the #343 to fix the memory consumption problem... It's a bigger change, but I'm planing to make it in this month.

**Comment by MetaDark** [comment]
Thanks, @rizsotto! Sorry about the duplicate issue. I saw #343, but didn't completely read it and thought it was unrelated.

**Comment by rizsotto** [comment]
This shall be fixed in version 3.0.7, which I just have released.

@MetaDark could you confirm that it fixes your issue?

**Comment by MetaDark** [comment]
@rizsotto Yes, it did! Thank you! It dramatically reduced memory consumption to about 50-60MB.

Here is the psrecord graph of intercept while building wine:
![psrecord](https://user-images.githubusercontent.com/382041/104680128-313fbb00-56bd-11eb-8a4e-5b273c94b7d4.png)
